lib/backend.js

```diff
@@ -568,6 +573,76 @@ Backend.prototype.getChannels = function(collection, id) {
   ];
 };
 
+Backend.prototype.getSnapshot = function(agent, index, id, version, timestamp, callback) {
+  var backend = this;
+  this._getSnapshot(agent, index, id, version, timestamp, function (error, snapshot) {
+    if (error) return callback(error);
+
+    var request = {
+      collection: index,
+      id: id,
+      version: snapshot.version,
+      timestamp: snapshot.timestamp,
+      snapshots: snapshot.data ? [snapshot.data] : [],
+      deleted: snapshot.deleted
+    };
+
+    backend.trigger(backend.MIDDLEWARE_ACTIONS.readSnapshots, agent, request, function (error) {
+      if (error) return callback(error);
+      callback(null, {
+        data: request.snapshots[0],
+        version: request.version,
+        timestamp: request.timestamp,
+        deleted: request.deleted
+      });
+    });
+  });
+};
+
+Backend.prototype._getSnapshot = function (agent, index, id, version, timestamp, callback) {
+  version = typeof version === 'number' ? version + 1 : null;
+  var options = { metadata: true };
+
+  this._getOps(agent, index, id, 0, version, options, function (error, ops) {
+    if (error) return callback(error);
+
+    var type;
+    var snapshot;
+    var deleted;
+    var op;
+
+    for (var index = 0; index < ops.length; index++) {
+      op = ops[index];
+
+      if (typeof timestamp === 'number' && op.m.ts > timestamp) {
+        op = ops[index - 1];
+        break;
+      }
+
+      if (op.create) {
+        type = types.map[op.create.type];
+        if (!type) return callback({ code: 4008, message: 'Unknown type' });
+        snapshot = op.create.data;
+        deleted = false;
+      } else if (op.del) {
+        snapshot = undefined;
+        deleted = true;
+      } else {
+        snapshot = type.apply(snapshot, op.op);
+      }
+    }
+
+    if (!snapshot && !deleted) return callback({ code: 4015, message: 'Document does not exist' });
```

Comment:
I don't think snapshot needs to be verified... the `create` and `apply` functions should guarantee that it's always correct for the given type. Additionally, in this context we're applying operations which were applied successfully before, which gives us even more confidence in the correctness of the result. Furthermore, `null`, `0` and `false` are all valid snapshots for ot-json.